cpp/src/cluster/detail/mst.cuh

```diff
@@ -117,6 +118,149 @@ void connect_knn_graph(
+ * @param[in] m number of rows in X
+ * @param[in] n number of columns in X
+ * @param[in] n_components number of components in color
+ * @param[inout] color the color labels array returned from the mst invocation
+ * @return updated MST edge list
+ */
+template <typename value_idx, typename value_t>
+void connect_knn_graph(
+  raft::resources const& handle,
+  const value_t* X,
+  raft::sparse::solver::Graph_COO<value_idx, value_idx, value_t>& msf,
+  size_t m,
+  size_t n,
+  int n_components,
+  value_idx* color,
+  cuvs::distance::DistanceType metric = cuvs::distance::DistanceType::L2SqrtExpanded)
+{
+  auto stream = raft::resource::get_cuda_stream(handle);
+
+  // Copy color array from device to host
+  std::vector<value_idx> h_color(m);
+  raft::copy(h_color.data(), color, m, stream);
+  raft::resource::sync_stream(handle, stream);
+
+  std::unordered_map<value_idx, value_idx> color_remap;
+  value_idx new_label = 0;
+
+  // Build remapping table so that colors are compact integers (i.e. consecutive colors)
+  for (size_t i = 0; i < m; ++i) {
+    if (color_remap.find(h_color[i]) == color_remap.end()) {
+      color_remap[h_color[i]] = new_label++;
+    }
+  }
+
+  for (size_t i = 0; i < m; ++i) {
+    h_color[i] = color_remap[h_color[i]];
+  }
+
+  // make key (color) : value (vector of ids that have that color)
+  std::unordered_map<value_idx, std::vector<value_idx>> component_map;
+  for (value_idx i = 0; i < static_cast<value_idx>(m); ++i) {
+    component_map[h_color[i]].push_back(i);
+  }
+
+  std::vector<std::tuple<value_idx, value_idx, value_t>> selected_edges;
+
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_int_distribution<> dis;
+
+  auto pairwise_dist = raft::make_device_matrix<value_t, int64_t>(handle, 1, 1);
+  auto data_u        = raft::make_device_matrix<value_t, int64_t>(handle, 1, n);
+  auto data_v        = raft::make_device_matrix<value_t, int64_t>(handle, 1, n);
+
+  // connect i-1 component and i component
+  for (int i = 1; i < n_components; ++i) {
+    value_idx color_a = i - 1;
+    value_idx color_b = i;
+
+    const auto& nodes_a = component_map[color_a];
+    const auto& nodes_b = component_map[color_b];
+
+    // Randomly pick a data index from each component
+    dis.param(std::uniform_int_distribution<>::param_type(0, nodes_a.size() - 1));
+    value_idx u = nodes_a[dis(gen)];
+
+    dis.param(std::uniform_int_distribution<>::param_type(0, nodes_b.size() - 1));
+    value_idx v = nodes_b[dis(gen)];
+
+    // calculate the distance between the two data vectors
+    raft::copy(data_u.data_handle(), X + u * n, n, stream);
+    raft::copy(data_v.data_handle(), X + v * n, n, stream);
+    cuvs::distance::pairwise_distance(handle,
+                                      raft::make_const_mdspan(data_u.view()),
+                                      raft::make_const_mdspan(data_v.view()),
+                                      pairwise_dist.view(),
+                                      metric);
```

Comment:
Based on your answer to my previous question, if using omp, I would also create new handles with streams picked from a pool for this call to saturate the GPU with concurrency (pairwise distance of 1 row at a time on the GPU will likely be slower than doing this on the CPU)